plugin/pkg/replacer/replacer.go

```diff
@@ -41,168 +56,247 @@ var labels = []string{
-		idxEnd := strings.Index(s[endOffset:], "}")
-		if idxEnd > -1 {
-			label := s[idxStart+2 : endOffset+idxEnd]
-
-			fm := metadata.ValueFunc(ctx, label)
-			replacement := EmptyValue
-			if fm != nil {
-				replacement = fm()
-			}
-
-			b.WriteString(s[:idxStart])
-			b.WriteString(replacement)
-			s = s[endOffset+idxEnd+1:]
-		} else {
-			break
-		}
-	}
-
-	b.WriteString(s)
-	return b.String()
-}
-
-func boolToString(b bool) string {
-	if b {
-		return "true"
+		return append(b, EmptyValue...)
+	default:
+		// TODO (CEV): consider panicking since this should be impossible
+		return append(b, EmptyValue...)
 	}
-	return "false"
 }
 
-// flagsToString checks all header flags and returns those
+// appendFlags checks all header flags and appends those
 // that are set as a string separated with commas
-func flagsToString(h dns.MsgHdr) string {
-	flags := make([]string, 7)
-	i := 0
-
+func appendFlags(b []byte, h dns.MsgHdr) []byte {
+	origLen := len(b)
 	if h.Response {
-		flags[i] = "qr"
-		i++
+		b = append(b, "qr,"...)
 	}
-
 	if h.Authoritative {
-		flags[i] = "aa"
-		i++
+		b = append(b, "aa,"...)
 	}
 	if h.Truncated {
-		flags[i] = "tc"
-		i++
+		b = append(b, "tc,"...)
 	}
 	if h.RecursionDesired {
-		flags[i] = "rd"
-		i++
+		b = append(b, "rd,"...)
 	}
 	if h.RecursionAvailable {
-		flags[i] = "ra"
-		i++
+		b = append(b, "ra,"...)
 	}
 	if h.Zero {
-		flags[i] = "z"
-		i++
+		b = append(b, "z,"...)
 	}
 	if h.AuthenticatedData {
-		flags[i] = "ad"
-		i++
+		b = append(b, "ad,"...)
 	}
 	if h.CheckingDisabled {
-		flags[i] = "cd"
-		i++
+		b = append(b, "cd,"...)
 	}
-	return strings.Join(flags[:i], ",")
+	if n := len(b); n > origLen {
+		return b[:n-1] // trim trailing ','
+	}
+	return b
 }
 
-// addrToRFC3986 will add brackets to the address if it is an IPv6 address.
-func addrToRFC3986(addr string) string {
-	if strings.Contains(addr, ":") {
-		return "[" + addr + "]"
+// appendAddrToRFC3986 will add brackets to the address if it is an IPv6 address.
+func appendAddrToRFC3986(b []byte, addr string) []byte {
+	if strings.IndexByte(addr, ':') != -1 {
+		b = append(b, '[')
+		b = append(b, addr...)
+		b = append(b, ']')
+	} else {
+		b = append(b, addr...)
 	}
-	return addr
+	return b
 }
 
+type nodeType int
+
 const (
-	headerReplacer = "{>"
-	labelReplacer  = "{/"
-	// EmptyValue is the default empty value.
-	EmptyValue = "-"
+	typeLabel    nodeType = iota // "{type}"
+	typeLiteral                  // "foo"
+	typeMetadata                 // "{/metadata}"
 )
+
+// A node represents a segment of a parsed format.  For example: "A {type}"
+// contains two nodes: "A " (literal); and "{type}" (label).
+type node struct {
+	value string // Literal value, label or metadata label
+	typ   nodeType
+}
+
+// A replacer is an ordered list of all the nodes in a format.
+type replacer []node
+
+func parseFormat(s string) replacer {
+	// Assume there is a literal between each label - its cheaper to over
+	// allocate once than allocate twice.
```

Comment:
... because `append` will double the capacity? And does it really matter that much here?

Reply:
Yes and no the extra allocs don't matter so long as parsing doesn't need to be very fast.  That said, I really don't blindly appending to a slice when the required size can be estimated.

Comparison of pre-allocate (current method) to appending.
```
$ benchcmp pre-allocate.txt append.txt
benchmark                   old ns/op     new ns/op     delta
BenchmarkParseFormat-12     527           918           +74.19%

benchmark                   old allocs     new allocs     delta
BenchmarkParseFormat-12     1              6              +500.00%

benchmark                   old bytes     new bytes     delta
BenchmarkParseFormat-12     704           1520          +115.91%
```